sys/mips/cheri/cheri_stack_machdep.c

```diff
@@ -0,0 +1,295 @@
+/*-
+ * SPDX-License-Identifier: BSD-2-Clause-FreeBSD
+ *
+ * Copyright (c) 2005 Antoine Brodin
+ * Copyright (c) 2020 Alfredo Mazzinghi
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/cdefs.h>
+__FBSDID("$FreeBSD$");
+
+#include <sys/types.h>
+#include <sys/systm.h>
+#include <sys/lock.h>
+#include <sys/mutex.h>
+#include <sys/param.h>
+#include <sys/proc.h>
+#include <sys/stack.h>
+
+#include <machine/mips_opcode.h>
+
+#include <machine/pcb.h>
+#include <machine/regnum.h>
+#include <machine/md_var.h>
+
+#include <machine/abi.h>
+
+#define	pc_in(pc, fn)							\
+	((pc) >= (uintptr_t)cheri_getbase(fn) &&			\
+	 (pc) < (uintptr_t)cheri_gettop(fn))
+
+static bool
+stack_addr_ok(struct thread *td, uintptr_t sp, u_register_t stack_pos)
+{
+	vm_pointer_t va = sp + stack_pos;
+
+	return (va >= td->td_kstack && va + sizeof(uintptr_t) <=
+	    td->td_kstack + td->td_kstack_pages * PAGE_SIZE);
+}
+
+static uintptr_t
+stack_register_fetch(uintptr_t sp, u_register_t stack_pos)
+{
+	uintptr_t * stack =
+	    ((uintptr_t *)sp + (size_t)stack_pos / sizeof(uintptr_t));
+
+	return (*stack);
+}
+
+static uintptr_t
+stack_fetch_ra(uintptr_t sp, u_register_t stack_pos)
+{
+	uintptr_t ra = stack_register_fetch(sp, stack_pos);
+
+	if (!ra)
+		return (ra);
+	/* Re-derive sentry capability */
+	if (cheri_getsealed(ra)) {
+		KASSERT(cheri_gettype(ra) == CHERI_OTYPE_SENTRY,
+		    ("Return address is not a SealedEntry capability %#p",
+		    (void *)ra));
```

Comment:
Feels like this should be a `return (NULL);` case rather than a (possibly nested) panic